pydantic_ai_slim/pydantic_ai/prompt.py

```diff
@@ -0,0 +1,184 @@
+    """Abstract base class for building tagged string representations of content.
+
+    This class provides the base functionality for converting various Python data types
+    into a tagged string format. It handles normalization and formatting of content
+    while leaving the specific string building implementation to subclasses.
+
+    Args:
+        tag: The root tag name to wrap the content with.
+        content: The content to be formatted. Supported types include:
+            - Basic types (str, int, float, bool)
+            - Dictionaries
+            - Lists, tuples, or other iterables
+            - Pydantic models
+            - Dataclasses
+
+    Attributes:
+        _content: Normalized and formatted content as a dictionary structure.
+
+    Raises:
+        TypeError: If content is of an unsupported type.
+    """
+
+    def __init__(self, tag: str, content: Any) -> None:
+        self._content: dict[str, Any] = self._format_content(tag, content)
+
+    @abstractmethod
+    def build(self) -> str: ...
+
+    def _format_content(self, tag: str, content: Any) -> dict[str, Any]:
+        root: dict[str, Any] = {}
+        normalized = _normalize_type(content)
+
+        if isinstance(normalized, dict):
+            root[tag] = {}
+            for key, value in normalized.items():
+                root[tag][key] = self._format_content(key, value)[key]
+        elif isinstance(normalized, list):
+            assert isinstance(normalized, list)
+            root[tag] = [self._format_content(tag, item)[tag] for item in normalized]
+        else:
+            root[tag] = normalized
+
+        return root
+
+
+@final
+class XMLTagBuilder(TagBuilder):
+    """Concrete implementation of TagBuilder that produces XML-like formatted output.
+
+    This class converts the normalized content into valid XML string representation.
+    It handles nested structures, lists, and basic types while ensuring proper XML formatting.
+
+    Examples:
+        >>> builder = XMLTagBuilder('user', {'name': 'John', 'age': 30})
+        >>> builder.build()
+        '<user><name>John</name><age>30</age></user>'
+
+        >>> builder = XMLTagBuilder('flags', [True, False])
+        >>> builder.build()
+        '<flags>true</flags><flags>false</flags>'
+
+        >>> builder = XMLTagBuilder('nested', {'user': {'details': {'active': True}}})
+        >>> builder.build()
+        '<nested><user><details><active>true</active></details></user></nested>'
+    """
+
+    def build(self) -> str:
+        return self._build_element(self._content)
+
+    def _build_element(self, content: dict[str, Any]) -> str:
+        result: list[str] = []
+
+        for tag, value in content.items():
+            if isinstance(value, dict):
+                nested_content = ''
+
+                for k, v in cast(dict[str, Any], value).items():
+                    nested_content += self._build_element({k: v})
+
+                result.append(f'<{tag}>{nested_content}</{tag}>')
```

Comment:
we need to support intending the XML with two spaces.

We also need to escape `</{tag>` within strings.